pkg/cache/cluster.go

```diff
@@ -864,6 +910,8 @@ func (c *clusterCache) sync() error {
 		return err
 	}
 
+	go c.processEvents()
```

Comment:
I see this in the `sync` function comment:

```go
// When this function exits, the cluster cache is up to date, and the appropriate resources are being watched for
// changes.
```

If I understand this change correctly (and the associated test changes), by processing these events in a goroutine, we're breaking the guarantee that `sync` will completely update the cluster cache. Is that correct?

Reply:
These changes do not break the guarantee that `sync` will completely update the cluster cache.
`sync` function populates the cluster cache when it's run.
 https://github.com/argoproj/gitops-engine/blob/a16e663ae6b5a5b0b904175e3672439c54ca7147/pkg/cache/cluster.go#L934

`processEvents` goroutine processes the future events that are received in `watchEvents` goroutine.
https://github.com/argoproj/gitops-engine/blob/a16e663ae6b5a5b0b904175e3672439c54ca7147/pkg/cache/cluster.go#L962

`watchEvents` goroutine watches for the events from k8s resource types. Once an event is received, it's processed.
https://github.com/argoproj/gitops-engine/blob/a16e663ae6b5a5b0b904175e3672439c54ca7147/pkg/cache/cluster.go#L706

The event is sent to the channel that is read in the `processEvents` goroutine, where the processing is done in bulk.
https://github.com/argoproj/gitops-engine/blob/a16e663ae6b5a5b0b904175e3672439c54ca7147/pkg/cache/cluster.go#L1299